crates/task-impls/src/quorum_proposal.rs

```diff
@@ -80,32 +66,184 @@ struct ProposalDependencyHandle<TYPES: NodeType> {
     sender: Sender<Arc<HotShotEvent<TYPES>>>,
 
     /// Reference to consensus. The replica will require a write lock on this.
-    #[allow(dead_code)]
     consensus: Arc<RwLock<Consensus<TYPES>>>,
+
+    /// Membership for Quorum Certs/votes
+    quorum_membership: Arc<TYPES::Membership>,
+
+    /// Our public key
+    public_key: TYPES::SignatureKey,
+
+    /// Our Private Key
+    private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
+
+    /// last Upgrade Certificate this node formed
+    upgrade_cert: Option<UpgradeCertificate<TYPES>>,
+}
+
+impl<TYPES: NodeType> ProposalDependencyHandle<TYPES> {
+    /// Checks if we have somehow initiated a proposal while not being the leader of the next
+    /// `view`.
+    fn check_leader(&self, view: TYPES::Time) -> Result<()> {
+        if self.quorum_membership.get_leader(view) != self.public_key {
+            // This is expected for view 1, so skipping the logging.
+            if view != TYPES::Time::new(1) {
+                bail!(
+                    "Somehow we formed a QC but are not the leader for the next view {:?}",
+                    view
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Find the leaf and state of the parent view to the view that we're currently proposing on.
+    fn find_parent_leaf_and_state(
+        consensus: &Consensus<TYPES>,
+    ) -> Result<(Leaf<TYPES>, Arc<TYPES::ValidatedState>)> {
+        let parent_view_number = &consensus.high_qc.get_view_number();
+
+        let Some(parent_view) = consensus.validated_state_map.get(parent_view_number) else {
+            // This should have been added by the replica?
+            bail!("Couldn't find parent view in state map, waiting for replica to see proposal\n parent view number: {}", **parent_view_number);
+        };
```

Comment:
could be `let parent_view = consensus.(...).context(format!("Couldn't find..", parent_view_number))?;`